Morph One

ryandialpad opened the pull request:
<!-- PULL REQUEST TEMPLATE -->

<!-- Please describe a summary of this PR. -->

#### Type of PR:

- [ ] Bugfix
- [x] Feature
- [ ] Refactor
- [ ] Code style update
- [ ] Build-related changes
- [ ] Test
- [ ] Documentation
- [ ] Other, please describe:

#### Breaking changes:

- [x] No
- [ ] Yes

### Details

<!-- Please describe the details of the PR. -->

Closes #90 

Adds Morph One Polymorphic relation & accompanying tests. Follow-up PR to add documentation.

The tests were created following the existing feature test pattern for the other relations.

ryandialpad commented on src/model/attributes/relations/MorphOne.ts:
```diff
@@ -0,0 +1,100 @@
+import { Relation, Dictionary } from './Relation'
+
+export class MorphOne extends Relation {
+  /**
+   * The field name that contains id of the parent model.
+   */
+  protected id: string
+
+  /**
+   * The field name that contains type of the parent model.
+   */
+  protected type: string
```
Following the API already established in Vuex ORM found in the following docs: https://github.com/vuex-orm/vuex-orm/blob/master/src/attributes/relations/MorphOne.ts

I personally would be open to renaming these two fields to make them a little more explicit so that the comments are a little less necessary.

Would you favour maintaining the previous interface or renaming these to something a bit more explicit?

kiaking replied:
Good point! This is internal stuff so I think it would be totally OK to rename it.

In Laravel, it's named `morphType` and `morphClass`. So maybe `morphId` and `morphType` might be easier to read?

ryandialpad replied:
Thanks for the feedback! I think `morphId` and `morphType` will be more clear for maintainers 👍  Going to add these changes

ryandialpad commented on src/model/attributes/relations/MorphOne.ts:
```diff
@@ -0,0 +1,100 @@
+   * Set the constraints for an eager load of the relation.
+   */
+  addEagerConstraints(query: Query, models: Collection): void {
+    query.where(this.type as any, this.parent.$entity())
+    query.whereIn(this.id as any, this.getKeys(models, this.localKey))
```
Quick explanation of the where constraints on the relation:

The type field should match the parent's entity and the id field should match the parent's local key

ryandialpad replied:
Removed `as any` as I don't think it is required for these queries.

kiaking replied:
Nice! Actually I think this `as any` is left here because before we were trying to type hint `where` query but we're not anymore, (figured out that it's nearly impossible). So I think we can remove this one entirely for other relations as well.

We can do that in another PR though.

ryandialpad replied:
Follow-up PR here: https://github.com/vuex-orm/vuex-orm-next/pull/97

ryandialpad commented:
I recommend squashing this PR when it is ready to be merged.

kiaking reviewed (commented):
@ryandialpad This is looking amazing! I've added a test case for `make` method to increase the coverage. But everything else looking good to me.

Once we resolve the discussion we have with the property names, I think we are ready to merge this 👍

ryandialpad commented:
> @ryandialpad This is looking amazing! I've added a test case for `make` method to increase the coverage. But everything else looking good to me.
> 
> Once we resolve the discussion we have with the property names, I think we are ready to merge this 👍

Thanks for the feedback! Updated `id` and `type` to `morphId` and `morphType` as discussed. Thanks for catching the missed test case 😃
